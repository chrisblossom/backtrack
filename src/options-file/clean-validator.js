/* @flow */

import { buildPath } from '../config/paths';
import { fileIsInsideDir } from '../utils/file-is-inside-dir';

import type { NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: $ReadOnlyArray<NormalizedClean>,
};

function cleanValidator({ value }: Args) {
    for (const arg of value) {
        for (const pattern of arg.del) {
            if (fileIsInsideDir(pattern, buildPath) === false) {
                throw new Error(
                    `del pattern '${pattern}' must be inside build directory`,
                );
            }
        }

        for (const dir of arg.makeDirs) {
            if (fileIsInsideDir(dir, buildPath) === false) {
                throw new Error(
                    `makeDirs '${dir}' must be inside build directory`,
                );
            }
        }
    }
}

export { cleanValidator };
