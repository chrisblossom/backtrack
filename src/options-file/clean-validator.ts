import path from 'path';
import { buildPath, rootPath } from '../config/paths';
import { fileIsInsideDirSync } from '../utils/file-is-inside-dir';
import { NormalizedClean } from '../types';

interface Args {
	value: NormalizedClean[];
}

function cleanValidator({ value }: Args): void {
	for (const arg of value) {
		for (const pattern of arg.del) {
			const resolvedPattern = path.resolve(buildPath, pattern);
			if (fileIsInsideDirSync(resolvedPattern, buildPath) === false) {
				throw new Error(
					`del pattern '${pattern}' must be inside build directory`,
				);
			}
		}

		for (const dir of arg.makeDirs) {
			if (fileIsInsideDirSync(dir, buildPath) === false) {
				throw new Error(
					`makeDirs '${dir}' must be inside build directory`,
				);
			}
		}

		for (const copy of arg.copy) {
			if (fileIsInsideDirSync(copy.src, rootPath) === false) {
				throw new Error(
					`copy.src '${copy.src}' must be inside project root directory`,
				);
			}

			if (fileIsInsideDirSync(copy.src, buildPath) === true) {
				throw new Error(
					`copy.src '${copy.src}' cannot be inside build directory`,
				);
			}

			if (fileIsInsideDirSync(copy.dest, buildPath) === false) {
				throw new Error(
					`copy.dest '${copy.dest}' must be inside build directory`,
				);
			}

			const isDuplicateWithHashMismatch = !!arg.copy.find((copyArg) => {
				return (
					copyArg.src === copy.src &&
					copyArg.dest === copy.dest &&
					copyArg.hash !== copy.hash
				);
			});

			if (isDuplicateWithHashMismatch) {
				throw new Error(
					`copy.src '${copy.src}' is duplicated with mismatched hash: true/false`,
				);
			}

			const isDuplicateWithSrcMismatch = !!arg.copy.find((copyArg) => {
				return copyArg.src !== copy.src && copyArg.dest === copy.dest;
			});

			if (isDuplicateWithSrcMismatch) {
				throw new Error(
					`copy.dest '${copy.dest}' is duplicated with mismatched sources`,
				);
			}
		}
	}
}

export { cleanValidator };
