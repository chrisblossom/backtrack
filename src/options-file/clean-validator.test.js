/* @flow */

import path from 'path';

const cleanValidator = (args) =>
    require('./clean-validator').cleanValidator(args);

describe('cleanValidator', () => {
    const cwd = process.cwd();
    const dir = path.resolve(__dirname, '../clean/__sandbox__/1/');

    beforeEach(() => {
        process.chdir(dir);
    });

    afterEach(() => {
        process.chdir(cwd);
    });

    test('fails with relative ../', () => {
        const value = [
            {
                del: ['../src/**'],
                makeDirs: [],
            },
        ];

        try {
            expect.hasAssertions();
            cleanValidator({ value });
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('fails with absolute inside rootPath', () => {
        const value = [
            {
                del: [path.resolve(dir, 'src/**')],
                makeDirs: [],
            },
        ];

        try {
            expect.hasAssertions();
            cleanValidator({ value });
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('fails with absolute', () => {
        process.chdir(__dirname);
        const value = [
            {
                del: [path.resolve(__dirname, '__sandbox__/should_fail')],
                makeDirs: [],
            },
        ];

        try {
            expect.hasAssertions();
            cleanValidator({ value });
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('fails with absolute + relative', () => {
        process.chdir(__dirname);
        const value = [
            {
                del: [path.resolve(__dirname, '__sandbox__/1/../dist/')],
                makeDirs: [],
            },
        ];

        try {
            expect.hasAssertions();
            cleanValidator({ value });
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('makeDirs fails', () => {
        process.chdir(__dirname);
        const value = [
            {
                del: [],
                makeDirs: [path.resolve(__dirname, '__sandbox__/1/../dist/')],
            },
        ];

        try {
            expect.hasAssertions();
            cleanValidator({ value });
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('passes with just del', () => {
        const value = [
            {
                del: ['*'],
                makeDirs: [],
            },
        ];

        const result = cleanValidator({ value });
        expect(result).toEqual(undefined);
    });

    test('passes with del glob', () => {
        const value = [
            {
                del: ['!.git'],
                makeDirs: [],
            },
        ];

        const result = cleanValidator({ value });
        expect(result).toEqual(undefined);
    });

    test('passes with just makeDirs', () => {
        const value = [
            {
                del: [],
                makeDirs: ['fake/dir/'],
            },
        ];

        const result = cleanValidator({ value });
        expect(result).toEqual(undefined);
    });
});
