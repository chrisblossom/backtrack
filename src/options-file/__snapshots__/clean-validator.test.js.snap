// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanValidator fails with absolute + relative 1`] = `[Error: del pattern '<PROJECT_ROOT>/__sandbox__/dist' must be inside build directory]`;

exports[`cleanValidator fails with absolute 1`] = `[Error: del pattern '<PROJECT_ROOT>/__sandbox__/should_fail' must be inside build directory]`;

exports[`cleanValidator fails with absolute inside rootPath 1`] = `[Error: del pattern '<PROJECT_ROOT>/src/**' must be inside build directory]`;

exports[`cleanValidator fails with relative ../ 1`] = `[Error: del pattern '../src/**' must be inside build directory]`;

exports[`cleanValidator makeDirs fails 1`] = `[Error: makeDirs '<PROJECT_ROOT>/__sandbox__/dist' must be inside build directory]`;
